src/decode/vec2d.rs

```diff
@@ -0,0 +1,74 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous `Vec`
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Vec<T>,
```

Comment:
Given that the `cols` and `rows` are immutable, this doesn't need a growable vector (with a capacity), but only a boxed slice `Box<[T]>`.

Reply:
Since `rows` and `cols` are variable length and not `const`, there's no way to construct the array without const generics or [placement `in`](https://github.com/rust-lang/rust/issues/22181)

Reply:
never mind, forgot about `into_boxed_slice`